File: mt5-websocket-client/src/utils/storage.js
const { BigQuery } = require('@google-cloud/bigquery');
const { logger } = require('./logger');

// Initialize BigQuery client
const bigquery = new BigQuery();

// Cache for created tables
const createdTables = new Set();

/**
 * Create BigQuery table if it doesn't exist
 * 
 * @param {string} datasetId BigQuery dataset ID 
 * @param {string} tableId BigQuery table ID
 */
async function ensureTableExists(datasetId, tableId) {
  // Check if we've already created this table in this session
  const tableKey = `${datasetId}.${tableId}`;
  if (createdTables.has(tableKey)) {
    return;
  }
  
  const dataset = bigquery.dataset(datasetId);
  const table = dataset.table(tableId);
  
  try {
    // Check if table exists
    const [exists] = await table.exists();
    
    if (!exists) {
      // Define table schema
      const schema = [
        { name: 'timestamp', type: 'TIMESTAMP' },
        { name: 'symbol', type: 'STRING' },
        { name: 'bid', type: 'FLOAT' },
        { name: 'ask', type: 'FLOAT' },
        { name: 'spread', type: 'FLOAT' }
      ];
      
      // Create table with time partitioning
      const options = {
        schema: schema,
        timePartitioning: {
          type: 'DAY',
          field: 'timestamp'
        }
      };
      
      logger.info(`Creating BigQuery table ${datasetId}.${tableId}`);
      await dataset.createTable(tableId, options);
      logger.info(`Created BigQuery table ${datasetId}.${tableId}`);
    }
    
    // Add to cache
    createdTables.add(tableKey);
    
  } catch (error) {
    logger.error(`Error creating BigQuery table: ${error.message}`);
    throw error;
  }
}

/**
 * Save data to BigQuery
 * 
 * @param {Object} data Record to insert
 * @param {string} datasetId BigQuery dataset ID
 * @param {string} tableId BigQuery table ID
 */
async function saveToBigQuery(data, datasetId, tableId) {
  try {
    // Ensure table exists
    await ensureTableExists(datasetId, tableId);
    
    // Format timestamp if it's an ISO string
    if (typeof data.timestamp === 'string') {
      data.timestamp = new Date(data.timestamp);
    }
    
    // Insert data
    const rows = [data];
    await bigquery.dataset(datasetId).table(tableId).insert(rows);
    
    return true;
  } catch (error) {
    logger.error(`BigQuery insert error: ${error.message}`);
    throw error;
  }
}

module.exports = {
  saveToBigQuery
};

File: mt5-websocket-client/src/utils/storage.py
import os
import csv
from google.cloud import storage
from google.cloud import bigquery


def save_to_csv(data, filename):
    """Save data to a CSV file
    
    Args:
        data (list): Row data to write [timestamp, symbol, bid, ask, spread]
        filename (str): Path to the CSV file
    """
    # Ensure the directory exists
    os.makedirs(os.path.dirname(filename), exist_ok=True)

    with open(filename, mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(data)


def save_to_gcs(data, bucket_name, destination_blob_name):
    """Save data to Google Cloud Storage
    
    Args:
        data (str): CSV formatted string to upload
        bucket_name (str): GCS bucket name
        destination_blob_name (str): Path within the bucket
    """
    client = storage.Client()
    bucket = client.bucket(bucket_name)
    blob = bucket.blob(destination_blob_name)

    blob.upload_from_string(data, content_type='text/csv')


def save_to_bigquery(data, dataset_id, table_id):
    """Save data to Google BigQuery
    
    Args:
        data (dict): Data record to insert
        dataset_id (str): BigQuery dataset ID
        table_id (str): BigQuery table ID
    """
    client = bigquery.Client()
    table_ref = client.dataset(dataset_id).table(table_id)
    
    # Define table schema if creating a new table
    schema = [
        bigquery.SchemaField("timestamp", "TIMESTAMP"),
        bigquery.SchemaField("symbol", "STRING"),
        bigquery.SchemaField("bid", "FLOAT"),
        bigquery.SchemaField("ask", "FLOAT"),
        bigquery.SchemaField("spread", "FLOAT"),
    ]
    
    # Create table if it doesn't exist
    try:
        client.get_table(table_ref)
    except Exception:
        table = bigquery.Table(table_ref, schema=schema)
        table.time_partitioning = bigquery.TimePartitioning(
            type_=bigquery.TimePartitioningType.DAY,
            field="timestamp"  # Partition by timestamp
        )
        client.create_table(table)
    
    # Insert data
    rows_to_insert = [data]
    errors = client.insert_rows_json(table_ref, rows_to_insert)
    
    if errors:
        raise Exception(f"BigQuery insert errors: {errors}")